File: sdks/js/src/verify.ts
import * as ed from "@noble/ed25519";

export type VerifyResult = { ok: boolean; reason?: string; kid?: string };

function _atobStd(s: string): string {
  if (typeof atob === "function") return atob(s);
  // Node fallback
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  return Buffer.from(s, "base64").toString("binary");
}

function b64ToBytes(s: string): Uint8Array {
  const bin = _atobStd(s);
  const out = new Uint8Array(bin.length);
  for (let i = 0; i < bin.length; i++) out[i] = bin.charCodeAt(i);
  return out;
}

function b64uToBytes(s: string): Uint8Array {
  let t = s.replace(/-/g, "+").replace(/_/g, "/");
  const pad = (4 - (t.length % 4)) % 4;
  t += "=".repeat(pad);
  return b64ToBytes(t);
}

export async function verifyEnvelope(envelope: any, jwksUrl: string): Promise<VerifyResult> {
  try {
    const opeB64 = envelope?.ope;
    if (!opeB64) return { ok: false, reason: "missing_ope" };
  const ope = JSON.parse(new TextDecoder().decode(b64ToBytes(opeB64)));
    const kid = ope?.kid;
    if (!kid) return { ok: false, reason: "missing_kid" };
    const oml = envelope?.oml_c_b64;
    if (!oml) return { ok: false, reason: "missing_oml_c" };
  // oml_c_b64 is standard base64
  const content = b64ToBytes(oml);
    const sig = b64uToBytes(ope?.sig_b64u || "");

    const res = await fetch(jwksUrl);
    if (!res.ok) return { ok: false, reason: "jwks_fetch_failed" };
    const jwks = await res.json();
    const k = jwks?.keys?.find((k: any) => k.kid === kid && k.kty === "OKP" && k.crv === "Ed25519" && k.x);
    if (!k) return { ok: false, reason: "kid_not_found", kid };
    const pub = b64uToBytes(k.x);

    const ok = await ed.verify(sig, content, pub);
    return ok ? { ok: true } : { ok: false, reason: "verify_failed" };
  } catch (e: any) {
    return { ok: false, reason: `verify_failed:${e?.name || typeof e}` };
  }
}

File: sdks/python/odin_protocol_sdk/verify.py
import json
import base64
import httpx
from dataclasses import dataclass
from typing import Optional
from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey


@dataclass
class VerifyResult:
    ok: bool
    reason: Optional[str] = None
    kid: Optional[str] = None


def _b64u_decode(s: str) -> bytes:
    return base64.urlsafe_b64decode(s + "==")


def _find_pub_from_jwks(jwks: dict, kid: str) -> Optional[bytes]:
    for k in jwks.get("keys", []):
        if k.get("kid") == kid and k.get("kty") == "OKP" and k.get("crv") == "Ed25519" and "x" in k:
            return _b64u_decode(k["x"])  # raw 32-byte public key
    return None


def verify_envelope(envelope: dict, jwks_url: str) -> VerifyResult:
    """Minimal OPE verify over inline content (oml_c_b64) using JWKS URL.
    Does not implement all gateway semantics; preview-only.
    """
    try:
        ope_b64 = envelope.get("ope")
        if not ope_b64:
            return VerifyResult(ok=False, reason="missing_ope")
        ope = json.loads(base64.b64decode(ope_b64).decode("utf-8"))
        kid = ope.get("kid")
        if not kid:
            return VerifyResult(ok=False, reason="missing_kid")
        oml_c_b64 = envelope.get("oml_c_b64")
        if not oml_c_b64:
            return VerifyResult(ok=False, reason="missing_oml_c")
        content = base64.b64decode(oml_c_b64)
        sig = base64.urlsafe_b64decode(ope.get("sig_b64u", "") + "==")
        # Fetch JWKS
        r = httpx.get(jwks_url, timeout=5.0)
        r.raise_for_status()
        jwks = r.json()
        pub = _find_pub_from_jwks(jwks, kid)
        if not pub:
            return VerifyResult(ok=False, reason="kid_not_found", kid=kid)
        Ed25519PublicKey.from_public_bytes(pub).verify(sig, content)
        return VerifyResult(ok=True, kid=kid)
    except Exception as e:
        return VerifyResult(ok=False, reason=f"verify_failed:{type(e).__name__}")
